### ISSUE TITLE
some ctrl-* keybindings only work with insert mode

### ISSUE THREAD

**Issue by diminutivesloop** [jeffreytse/zsh-vi-mode#195]
## General information

- Terminal program: Terminal 2.12.7
- Operating system: Mac OS 12.6.1
- ZSH framework: oh-my-zsh
- ZSH version: 5.8.1
- ZVM version: 0.9.0

## Basic examination

- [x] I have read through the [README](https://github.com/jeffreytse/zsh-vi-mode) page
- [x] I have the latest version of zsh-vi-mode
- [ ] I have tested with another terminal program 

## Problem description

Some ctrl-* keybindings only work in insert mode. Specifically I'm using the per-directory-history zsh plugin and I cannot use ctrl-g to switch between global and directory history when in normal mode even though other keyboard commands such as ctrl-c work in both modes. I'm initializing per-directory-history via `zvm_after_init` and I tried disabling lazy loading via `ZVM_LAZY_KEYBINDINGS=false` but this didn't fix the issue.

## Reproduction steps
N/A

## Expected behavior

Any ctrl-* keybindings that are not assigned to this plugin should work as expected in all input modes.

**Comment by cohml** [comment]
Big bump for this one please!!!

**Comment by jeffreytse** [comment]
Hi @diminutivesloop 

Thanks for your reporting, I want to confirm the bellowing items with you firstly:

1. Is it [this plugin](https://github.com/jimhester/per-directory-history)?
2. Are you sure without enabling zsh-vi-mode, this plugin can work under vi normal mode?

Thanks & Regards

**Comment by cohml** [comment]
I'm not @diminutivesloop, but I can be another data point.

I use [this plugin](https://github.com/ohmyzsh/ohmyzsh/tree/master/plugins/per-directory-history), which is the same as [the one you linked](https://github.com/jimhester/per-directory-history), except that I install it via Oh-My-Zsh.

I have confirmed that when I remove `zsh-vi-mode` from the OMZ plugins list and reload, `CTRL-G` toggles my history as intended. When I bring `zsh-vi-mode` back, toggling no longer works.

**Comment by cohml** [comment]
I'm not @diminutivesloop, but I can be another data point.

I use [this plugin](https://github.com/ohmyzsh/ohmyzsh/tree/master/plugins/per-directory-history), which is the same as [the one you linked](https://github.com/jimhester/per-directory-history), except that I install it via Oh-My-Zsh.

With `per-directory-history` enabled, here's what I've observed:

- When I enable `zsh-vi-mode`, `CTRL-G` stops working.
- When I disable `zsh-vi-mode` but enable `bindkey -v`, `CTRL-G` stops working.
- When I disable `zsh-vi-mode` and disable `bindkey -v`, `CTRL-G` works.

**Comment by jeffreytse** [comment]
@diminutivesloop Have you tried to source the plugin as below?

```sh
zvm_after_init() {
  source <your_plugin_location>/per-directory-history.plugin.zsh
}
```

**Comment by cohml** [comment]
Hey they fixes it for me!

I know your README covers this a bit. For some reason it hadn't occurred to me that `zvm_after_init` could also work with zsh plugins. But in retrospect that makes perfect sense. [The README even mentioned `zsh-autocomplete` specifically.](https://github.com/jeffreytse/zsh-vi-mode#:~:text=e.%20fzf%2C-,zsh%2Dautocomplete,-%2C%20etc.)%20to%20fail) I think your bases are all already covered here 😅

Sorry for the confusion and thanks for the help! I vote to close this issue.

**Comment by jeffreytse** [comment]
@diminutivesloop You're welcome, and hope this plugin can accelerate your daily work. And also welcome to star this project for further updates in the future. : )

**Comment by cohml** [comment]
I'm still not @diminutivesloop lol, so I'd wait for him to respond before closing.

But I will star for sure. This project is great and you are a very responsive dev. Thanks again!

**Comment by jeffreytse** [comment]
@cohml You're welcome, welcome to submit your issues whenever you meet. : )

**Comment by diminutivesloop** [comment]
> Hi @diminutivesloop
> 
> Thanks for your reporting, I want to confirm the bellowing items with you firstly:
> 
>     1. Is it [this plugin](https://github.com/jimhester/per-directory-history)?
> 
>     2. Are you sure, without enabling zsh-vi-mode, this plugin can work under vi normal mode?
> 
> 
> Thanks & Regards

Yes that's the plugin I'm referring to. And I did just test with the regular vi-mode plugin and the issue manifests in that case as well.

**Comment by cohml** [comment]
@diminutivesloop Try [this](https://github.com/jeffreytse/zsh-vi-mode/issues/195#issuecomment-1645681984. I had the same issue as you, and this solution seems to have fixed it.

**Comment by jeffreytse** [comment]
@diminutivesloop Could you try as @cohml said, and see if your issue could be settled down?

**Comment by diminutivesloop** [comment]
@jeffreytse That appears to work. I tried the same thing before with no success, but that may have been before I switched to installing the `HEAD` version of this plugin to address another issue.

**Comment by wookayin** [comment]
Having to source other plugins `zvm_after_init` is just too hacky (as in #24) and quite painful. Can we fix the root cause of key conflicts? This interferes with almost every other zsh plugins, and depending on how zsh plugins are loaded (e.g. through a plugin manager) one can't really put everything in this hacky `zvm_after_init` function.

**Comment by cohml** [comment]
@wookayin

> Having to source other plugins `zvm_after_init` is just too hacky (as in #24) and quite painful. Can we fix the root cause of key conflicts? This interferes with almost every other zsh plugins, and depending on how zsh plugins are loaded (e.g. through a plugin manager) one can't really put everything in this hacky `zvm_after_init` function.

100% agree. I've long since stopped using this plugin for that exact reason.

ZVM is really great: definitely the best vi mode implementation, and @jeffreytse is a very responsive dev; I did not abandon it lightly. But the feature set it brings does not outweigh the pain of integrating it with other plugins. It just clobbers too many things and often in non-obvious ways.

The only sustainable way forward is to rethink ZVM's core design and fix it at the source rather than stacking bandaids on bandaids with `zvm_after_init`.

**Comment by jeffreytse** [comment]
@wookayin @cohml Thanks for reclaiming this issue, currently it has a way to change the default loading strategy to initialize this plugin at sourcing stage, the default way of postponing initialization is to prevent other annoying problems when I found in the development, this is some kind of trade-offs.

💯 I'd like to refactor this plugin when I have more time, and thank you so much for you all supports on this project. And these days, I'm working on another new project [grimoire](https://github.com/jeffreytse/grimoire), which is to democratize best practices across 27 domains, make your AI agent practicing it rather than knowing it, hope you guys like my new work. : )
